### ISSUE TITLE
Cloud Run + Cloud SQL hanging issue

### ISSUE THREAD

**Issue by davidmartos96** [GoogleCloudPlatform/functions-framework-dart#302]
Hello, we have been migrating our infra to Cloud Run + Cloud SQL Postgres + Shelf, and while everything worked great with HTTP handlers, once we've added Cloud SQL, we've encountered a roadblock.
It seems like when the Cloud SQL is opened via the Unix socket on Cloud Run, the HTTP server hangs, and connections afterwards don't work.
Locally, if the app uses the Unix socket created by the Cloud SQL proxy, it works ok. It's just in the Cloud Run environment where we are experiencing this issue. 

We are also unsure where this issue belongs, because Unix socket connections work with the Postgres library. Could it be possible that is an issue in the Dart SDK?

Here is a very simple demo that deploys a Cloud Run function that uses a Cloud SQL instance. To test the issue one only needs to connect to the root of the host and then refresh the page.

https://github.com/davidmartos96/cloud-run-dart-pg-issue

**Comment by kevmoo** [comment]
CC @mit-mit – I'm guessing this is not specific to this framework, and likely something deeper in pkg:shelf or dart:io

Will dig in here!

**Comment by kevmoo** [comment]
Could you get a repro of this working with docker-compose? (With Dart + postgres?)

Would make debugging a LOT easier!

**Comment by davidmartos96** [comment]
> Could you get a repro of this working with docker-compose? (With Dart + postgres?)
> 
> Would make debugging a LOT easier!

You mean working with a regular Postgres database on Docker with Unix sockets? I can try to do that.

**Comment by kevmoo** [comment]
Yeah – and if you included a `docker compose` setup yaml in the repo, it makes reproducing on our side a LOT easier!

**Comment by davidmartos96** [comment]
@kevmoo I've updated the sample with a docker compose ready to run the app with a local Postgres. As you'll see, the socket connection works correctly locally.

**Comment by kevmoo** [comment]
I'd be surprised if unix socket works on run at all. You're saying it DOES work?

Have you tried TCP? Does that work?

**Comment by kevmoo** [comment]
Are you using Cloud SQL? 

Have you looked at https://cloud.google.com/sql/docs/postgres/connect-run#connecting_to ?

**Comment by davidmartos96** [comment]
@kevmoo Yes, Unix sockets work correctly, but only locally 😅
TCP works just fine both locally and on Cloud Run. The problem is that in the Cloud Run environment the IP is unknown, so we cannot add an authorized network for it in Cloud SQL. We would need to add 0.0.0.0/0 to use TCP from Cloud Run.

Yes, we looked at the provided docs, and that's how we've got to connect it. The issue is just in the second connection. Our Cloud SQL uses public IP, so the docs state to use the Unix socket connection mounted in the /cloudsql path.

**Comment by davidmartos96** [comment]
We've also tried deploying the node.js demo, which connects to the same socket path, and it works. It seems to be something related to Dart.

**Comment by davidmartos96** [comment]
Hey @kevmoo, any updates on this? Like some other issue I could track?

**Comment by marcusmonteirodesouza** [comment]
I think I might be experiencing the same problem. When I deploy [this app](https://github.com/marcusmonteirodesouza/dart-shelf-realworld-example-app) to Cloud Run (via [terraform](https://github.com/marcusmonteirodesouza/dart-shelf-realworld-example-app/blob/master/deploy/gcp/terraform/main.tf)) and then make a HTTP request to it, in the majority of times the request hangs on indefinitely until it returns `504`, and sometimes it returns OK. I even opened an  [issue](https://issuetracker.google.com/issues/220618415) for that because it is puzzling.

**Comment by brianquinlan** [comment]
Is there currently an owner for this issue? If not, I'll assign it (and the related https://github.com/dart-lang/sdk/issues/47899) to myself.

**Comment by kevmoo** [comment]
@brianquinlan Please! Happy to help here. I don't have the tech chops to help figure out what's happening.

Other folks here, I have a question: have you ever tried to containerize your app locally and then connect to Cloud SQL?

Trying to figure out if this is a Docker/container thing or a Cloud Run thing...

**Comment by davidmartos96** [comment]
@kevmoo When I encountered with this issue I tried connecting via Unix socket through the Cloud SQL Proxy (https://cloud.google.com/sql/docs/postgres/sql-proxy) and the socket connection through Dart was working fine, but I wasn't able to containerize the proxy.

From the tests I did my guess is that it could be something in Dart io or Cloud Run, and not a Postgres library issue.

**Comment by marcusmonteirodesouza** [comment]
@kevmoo I created a branch to test this https://github.com/marcusmonteirodesouza/dart-shelf-realworld-example-app/tree/test/cloudsql-proxy-containerized

I made some requests to my local endpoint and they all returned `OK`. I checked and the data was saved in my CloudSQL database. So a local containerized app connected to CloudSQL Proxy seems to be working.

**Comment by brianquinlan** [comment]
I've assigned the Dart issue (https://github.com/dart-lang/sdk/issues/47899) to myself but I probably won't have time to take a look at this until next week.

**Comment by brianquinlan** [comment]
I'm pretty sure that this is a gVisor issue.

I've filed a bug with the gVisor team: https://github.com/google/gvisor/issues/7331

@kevmoo The workaround is to use the gen2 execution environment, see [selecting an execution environment](https://cloud.google.com/run/docs/configuring/execution-environments).

Do we want to update our documentation (e.g. https://github.com/GoogleCloudPlatform/functions-framework-dart/blob/main/docs/quickstarts/03-quickstart-cloudrun.md) to suggest using the gen2 execution environment?

**Comment by brianquinlan** [comment]
@kevmoo I can change the Cloud Run quickstart. Do you know of any other documents that need to be updated?

**Comment by brianquinlan** [comment]
@kevmoo Also, are you an admin for this repo? If so, could you give me permission to modify issues?

**Comment by kevmoo** [comment]
I don't think so. We found the root cause.

**Comment by AlexKenbo** [comment]
> There are more details here: [dart-lang/sdk#47899](https://github.com/dart-lang/sdk/issues/47899)
> 
> But my current hypothesis is that the gVisor containerization used in the gen1 Cloud Run execution environment implements a syscall differently than Dart expects.
> 
> You can work around that by using the gen2 execution environment, see [selecting an execution environment](https://cloud.google.com/run/docs/configuring/execution-environments).
> 
> I can no longer reproduce the issue if I change this repo: https://github.com/davidmartos96/cloud-run-dart-pg-issue
> 
> with:
> 
> ```diff
> diff --git a/Makefile b/Makefile
> 		--add-cloudsql-instances=$(CLOUD_SQL_CONNECTION_NAME) \
> 		--set-env-vars=DB_USER=$(DB_USER),DB_PASS=$(DB_PASS),DB_NAME=$(DB_NAME),CLOUD_SQL_CONNECTION_NAME=$(CLOUD_SQL_CONNECTION_NAME) \
> +		--execution-environment=gen2
> ```
> 
> My next step is to try to reproduce this using gVisor directly to see if this is a bug in Dart or gVisor.


Thank you it helped only this (gen2)
<img width="538" alt="Снимок экрана 2022-11-14 в 21 39 23" src="https://user-images.githubusercontent.com/14903011/201739947-2dd6ce7f-8483-4b3a-b270-2569b58695e8.png">
